qiskit_nature/circuit/library/initial_states/hartree_fock.py

```diff
@@ -39,7 +40,14 @@ def __init__(
             num_particles: The number of particles as a tuple storing the number of alpha- and
                            beta-spin electrons in the first and second number, respectively.
             qubit_converter: a QubitConverter instance.
+
+        Raises:
+            ValueError: If qubit_converter contains BravyiKitaevSuperFastMapper.
         """
+        if isinstance(qubit_converter.mapper, BravyiKitaevSuperFastMapper):
+            raise ValueError(
```

Comment:
Not sure about the `ValueError`. I would rather have it be a `QiskitNatureError`

Reply:
Ok. I just copied other errors in qiskit-nature that used `ValueError` for an "unsupported" value.

Reply:
I also see `TypeError` for the same thing in qiskit-nature:
```
qiskit_nature/transformers/second_quantization/electronic/active_space_transformer.py
427:            raise TypeError(f"{type(prop)} is an unsupported Property-type for this Transformer!"
```

Reply:
Huh interesting. We should maybe try to unify that as part of a non-backportable PR..
@woodsp-ibm @manoelmarques @pbark Do you have any thoughts on this?

Reply:
The python docs suggest this is the one: https://docs.python.org/3/library/exceptions.html#TypeError
But, I have often seem software raise errors that name the project, eg `QiskitNatureError`. I am currently searching for discussions on why to choose one or the other.

Reply:
I think generally its better to raise one of the standard exceptions when it applies, as would be the case for type or value checking if its invalid as its more specific rather than using the nature custom one. For me the QiskitNatureError is more a general catch-all for others that don't fit - though arguably one could use the generic RuntimeError and we need not have a custom one. I think these more generic custom ones were to be consistent with Terra that defined/used its own custom one; it does have more specific subclasses of that for certain error categorization.

Reply:
Here is a guide that discusses some designs with references to real-world projects. There are a couple of strategies that do both at the same time: tell what library the exception occurred in and what kind of error.
https://julien.danjou.info/python-exceptions-guide/

One strategy is to follow what is already in qiskit-nature by subclassing `QiskitNatureError`:
https://github.com/Qiskit/qiskit-nature/blob/d6a4db83942cd55c3e23cfbf98497bbc41779594/qiskit_nature/exceptions.py#L24-L27

It might be `UnsupportedTypeError`. But, the python docs say that `TypeError` means exactly that a type is unsupported. So perhaps just `TypeError(QiskitNatureError)`.

Reply:
> I think generally its better to raise one of the standard exceptions when it applies

The arguments in the links above say that the reason to have something like `class TypeError(QiskitNatureError)` is so that the consumer can see which library the error occurred in (And maybe facilitate the consumer's handling of the errors).  Even, if this is a reasonable approach, I wonder the extent to which qiskit-nature is a library, i.e. called from higher level code, or an application, i.e. with classes instantiated and methods called in a jupyter notebook. The idea is that an error handling hierarchy is useful for a library that may be used in conjunction with many others in an application. Also, how important is it to get correct right now? If it changed would we have to deprecate it? I mean, someone could write code that catches this error. If we change the error, that code would break.

But, if it is just between `QiskitNatureError` and `TypeError`, then as @woodsp-ibm says, the latter is clearly better. In fact, as I read about it, I don't find anyone recommending a design that has just one error class for a whole library. (I looked in my favorite/only Python book, "Fluent Python", but it doesn't address exceptions at all)

Reply:
I'd agree that is the error is distinct where some application code can reasonably act on the error differently then it makes sense - I am just not sure a subclass of type error would ever do that. A consumer can see where the error in the stack trace, and hopefully its clear enough from that - given our naming I think it is.

Reply:
> consumer can see where the error in the stack trace

Yes, that's important; it's why I don't understand the rationale entirely. The only thing I can think of is the same as what the author of the post above says:
> It then makes it easy to use except ShoeError when doing anything with that piece of code related to shoes. For example, Django does not do that for some of its exceptions, making it hard to catch "any exception raised by Django".

So I have a function that calls into qiskit and might go through nature and other libraries or structures. When catching errors, I might want to begin by catching "any exception raised by qiskit-nature". But, I don't have, and the author did not give, a concrete example.

Reply:
Subclassing seems like added complexity unless we can articulate a clear use case. The Python docs themselves talk about a different reason for custom exceptions. The exception is really not described by any of the standard ones and/or it is useful to include some additional attributes for debugging. But, in this case, we have precisely a `TypeError`.

Reply:
Another example, JAX: They raise both builtin and custom errors (although some builtin errors are reworked by decorators). They subclass `TypeError`. So instead of `TypeError(QiskitNatureError)`, they do `QiskitNatureTypeError(TypeError)`.  Then they further subclass these. Why they have to be `TypeError` is not clear. But, they do clearly have a reason. They signal specific types of errors, and include information, such a a link to a webpage.  So JAX seems to raise builtin when they can, and custom otherwise.

We might want to consider having a more user-friendly "frictionless" exception setup, with webpages. Or maybe `IncompatibleDriver` classes or something like that. But, we don't have any of that in place yet.

Reply:
From what I gather here it seems that this is a longer discussion. I see you already opened a separate issue for this (#505). Shall we use a `TypeError` (as you said is suggested by the Python docs) for the time being?